services/telemetry/base.go

```diff
@@ -0,0 +1,154 @@
+package telemetry
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+)
+
+// Telemetry exported and unexported fields
+type Service struct {
+	Config
+	ticker     *time.Ticker
+	lastStatus string
+	stopChan   chan bool
+	wg         sync.WaitGroup
+	isRunning  bool
+	lock       sync.Mutex
+}
+
+// Telemetry config
+type Config struct {
+	URL      string `yaml:"url'`
+	UUID     string `yaml:"uuid"`
+	Interval int    `yaml:"interval"`
+}
+
+const (
+	defaultURL      = "https://v.percona.com"
+	defaultUUID     = ""
+	defaultinterval = 24 * 60 * 60
+)
+
+var (
+	stat     = os.Stat
+	readFile = ioutil.ReadFile
+	output   = commandOutput
+	Version  = "1.3.0"
+)
+
+// NewService creates a new telemetry service given a configuration
+func NewService(config *Config) (*Service, error) {
+	service := &Service{
+		Config: *config,
+	}
+	return service, nil
+}
+
+// Start sets a new ticker to collect data every Config.Interval seconds.
+func (s *Service) Start() {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if s.isRunning {
+		return
+	}
+	s.stopChan = make(chan bool)
+	s.isRunning = true
```

Comment:
A lot of code there can be simplified with context. Something like:
```go
// Run runs Call home service until context is done.
func (s *Service) Run(ctx context.Context) {
  for {
    select {
      // select on ticker and ctx.Done()
    }

    // do work
  }
}
```

Chans, bool var, `Stop()` method can be dropped.

Reply:
I can change it if you want but I prefer to have a way to manually stop the service and to acknowledge it was really stopped.

Reply:
Context with sync.WaitGroup allows you to exactly that with less code and with the same interface – cancel the context to stop a running thing.